packages/puppeteer-core/src/cdp/IsolatedWorld.ts

```diff
@@ -65,36 +75,45 @@ export class IsolatedWorld extends Realm {
   }
 
   setContext(context: ExecutionContext): void {
-    const existingContext = this.#context.value();
-    if (existingContext instanceof ExecutionContext) {
-      existingContext[disposeSymbol]();
-    }
+    this.#context?.[disposeSymbol]();
     context.once('disposed', () => {
-      // The message has to match the CDP message expected by the WaitTask class.
-      this.#context?.reject(new Error('Execution context was destroyed'));
-      this.#context = Deferred.create();
+      this.#context = undefined;
       if ('clearDocumentHandle' in this.#frameOrWorker) {
         this.#frameOrWorker.clearDocumentHandle();
       }
     });
-    this.#context.resolve(context);
+    this.#context = context;
+    this.#emitter.emit('context', context);
     void this.taskManager.rerunAll();
   }
 
   hasContext(): boolean {
-    return this.#context.resolved();
+    return !!this.#context;
   }
 
-  #executionContext(): Promise<ExecutionContext> {
+  #executionContext(): ExecutionContext|undefined {
     if (this.disposed) {
       throw new Error(
         `Execution context is not available in detached frame "${this.environment.url()}" (are you trying to evaluate?)`
       );
     }
-    if (this.#context === null) {
-      throw new Error(`Execution content promise is missing`);
-    }
-    return this.#context.valueOrThrow();
+    return this.#context;
+  }
+
+  async #waitForExecutionContext(): Promise<ExecutionContext> {
+    const result = await firstValueFrom(
```

Comment:
We can do something like bellow so that we don't need to check everywhere if the context already exits
```suggestion
     merge(
     fromEmitterEvent(),
     from(this.#context)
     ).pipe(
     filter(Boolean),
     raceWith()
     )
```

Reply:
Nope, since await messed up the relative execution w.r.t to the caller code :D so we want to at least in most cases be synchronous :D